src/riak_client.erl

```diff
@@ -398,20 +399,45 @@ delete(Bucket,Key,Options,Timeout,{?MODULE, [Node, _ClientId]}=THIS) when is_lis
         true ->
             consistent_delete(Bucket, Key, Options, Timeout, THIS);
         false ->
-            normal_delete(Bucket, Key, Options, Timeout, THIS);
+            maybe_normal_delete(Bucket, Key, Options, Timeout, THIS);
         {error,_}=Err ->
             Err
     end;
 delete(Bucket,Key,RW,Timeout,{?MODULE, [_Node, _ClientId]}=THIS) ->
     delete(Bucket,Key,[{rw, RW}], Timeout, THIS).
 
+%% If this is a write-once bucket, delete without read-before-delete
+%% (write_once_delete).  Else do a normal delete.
+
+maybe_normal_delete(Bucket, Key, Options, Timeout, {?MODULE, [Node, ClientId]}) ->
+    case write_once(Node, Bucket) of
+        true ->
+	    write_once_delete(Bucket, Key, Options, Timeout, {?MODULE, [Node, ClientId]});
+        false ->
+	    normal_delete(Bucket, Key, Options, Timeout, {?MODULE, [Node, ClientId]});
+        {error,_}=Err ->
+            Err
+    end.
+
+%% This version is exported, since it is called directly from
+%% riak_kv_ts_api:delete_data/4
+
+write_once_delete(Bucket, Key, Options, {?MODULE, [Node, ClientId]}) ->
+    write_once_delete(Bucket, Key, Options, ?DEFAULT_TIMEOUT, {?MODULE, [Node, ClientId]}).
+
+%% Write-once get path ignores vclocks.  We pass in an empty vclock
+%% (vclock:fresh()) so that no read-before-delete is triggered in
+%% riak_kv_delete:delete/8
+
+write_once_delete(Bucket, Key, Options, Timeout, {?MODULE, [Node, ClientId]}) ->
+    normal_delete_vclock(Bucket, Key, vclock:fresh(), Options, Timeout, {?MODULE, [Node, ClientId]}).
```

Comment:
You're absolutely right -- I missed that we had already changed this logic in the TS branches, from: 

```
case crypto:hash(sha, term_to_binary(OldObject)) =< crypto:hash(sha, term_to_binary(NewObject)) of
        true ->
            OldObject;
        _ ->
            NewObject
    end.
```

to 

``` 
pick_last_w1c_timestamp({OldObject, vclock:get_timestamp(<<0:8>>, OldObject#r_object.vclock)},
                            {NewObject, vclock:get_timestamp(<<0:8>>, NewObject#r_object.vclock)}).
```
(I thought that change had only been made in the KV branches).

Given that, I think that if we use `riak_kv_w1c_worker:w1c_vclock/1` in place of `vclock:fresh()`, it should have the correct behavior: a tombstone with a later timestamp than a repair will win, but a near-simultaneous update of the value with a later timestamp than the tombstone will win over the tombstone.